protocol.md

```diff
@@ -568,6 +568,127 @@ Tus-Resumable: 1.0.0
 Upload-Offset: 11
 ```
 
+### Partial Checksum
+
+This extension allows Clients that use the Checksum extension to verify and confirm the
+integrity of chunks that were not received in full by the Server - thereby avoiding that
+intact data needs to be sent again.
+
+Servers that support this extension MUST add `partial-checksum` to the `Tus-Extension` header.
+
+If a Server receives an incomplete `PATCH` or `POST` request with `Upload-Checksum` header, it
+MAY store the incompletely received chunk's data (hereafter referred to as *partial data*) in
+a separate location instead of discarding it.
+
+If the Server receives a `HEAD` request on the upload thereafter, it MAY compute a checksum
+on the partial data and then provide the result as well as the range of the partial data within
+the upload through the `Upload-Partial-Checksum` and `Upload-Partial-Checksum-Range`
+response headers.
+
+A Client can then use the range returned by the Server, compute its own checksum for the 
+same range and compare it against the checksum returned by the Server.
+
+If the checksums are identical, the Client SHOULD confirm the partial data's integrity to the
+Server by sending the checksum and range in the `Upload-Partial-Checksum` and
+`Upload-Partial-Checksum-Range` headers of the next `PATCH` request. The `Upload-Offset`
+header of the `PATCH` request MUST take the confirmed range into account and represent
+the actual offset from where the upload should be resumed.
+
+If the checksums don't match, the Client MUST continue the upload from the `Upload-Offset`
+returned by the `HEAD` request, at which point the Server MAY discard any partial data it may
+still be holding.
```

Comment:
I think this section here is very clear. Nicely done! Would this be correct from a servers perspective just to clarify that I understand this correctly? The pretense is that the client started a PATCH request with a Upload-Checksum header and then disconnected:

* The HEAD request will contain `Upload-Partial-Checksum` and the `Upload-Partial-Checksum-Range` headers.
* In the next PATCH request, if the `Upload-Partial-Checksum` matches the bytes for the file specified in `Upload-Partial-Checksum-Range` then the new Upload-Offset is the upload offset before the disconnected PATCH request + the range of the range header. If the `Upload-Partial-Checksum` header is not the correct checksum (or is missing) then the Upload-Offset is the offset from before the disconnected PATCH request and the server could safely remove the partial data.
* _Do we wish to just discard the partial data if the `Upload-Partial-Checksum` is invalid or do we wish to return an error to the client?_  Scratch this, it is answered below. We will return a 409 Conflict.

Reply:
> * The HEAD request will contain `Upload-Partial-Checksum` and the `Upload-Partial-Checksum-Range` headers.

If you mean that these headers would be contained in the response to the HEAD request, then: yes, that's exactly how it is meant to work.

Reply:
Yes sorry! Obviously meant the HEAD response :)